README.md
# TBN

This projects aims to provide music entertainment for toddlers.

The idea is that children will get a WiFi connected box and some objects (e.g. PlayMobil figures) that are equipped with NFC stickers. Whenever the child is putting such an object on the box, the box will start to play audio.
The NFC to audio URL translation can be configured on a server in the Azure cloud.
So a PlayMobil knight might make the box play a story about knights, or a picture with grandpa may play his voice telling a small story.

There are couple of commercial solutions for this, but this project aims to create an open platform.

## Components

If you want to build your own jukebox you will need the items from the [shopping list](AudioBoxConrad.pdf).

### Raspberry Pi

The Raspberry Pi will be connected to an NFC/RFID-Card reader and a speaker module.

### Server

The Server component will be run on Azure and provides a service that resolves NFC tags to download/streaming URLs.

### WebClient

The web client will allow users to configure their Raspberry Pi and add NFC tags or upload audio files.

## Additional features

### Local storage

Users may want to keep audio files in their local network. The web client will allow to configure such links to local NAS.

### Business times

Nobody wants their children to stay awake all night to play with this box. The website will allow to configure times where the system is inactive.

## Installation

### Preparing the SD-Card

#### Flash operating system on SD-Card

* Download latest [Raspbian image with Desktop](https://www.raspberrypi.org/downloads/raspbian/)
* Use [Etcher](https://etcher.io/) to flash the image onto your SD-Card

#### Activate WiFi

* Create a file called `wpa_supplicant.conf` on the SD-Card and put the following content in:

```ini
ctrl_interface=DIR=/var/run/wpa_supplicant GROUP=netdev
update_config=1
country=<<Your_ISO-3166-1_two-letter_country_code>>

network={
    ssid="<<Your_SSID>>"
    psk="<<Your_PSK>>"
    key_mgmt=WPA-PSK
}
```

* Replace <<Your_ISO-3166-1_two-letter_country_code>> with your ISO Country Code (such as DE for Germany), <<Your_SSID>> with your wireless access point name and <<Your_PSK>> with your wifi password.
* ATTENTION: The file must be formatted with Linux style line endings.
* More details can be found on [stackexchange](https://raspberrypi.stackexchange.com/a/57023).

#### Activate SSH

* Create a file called `ssh` (no ending) on the SD-Card. It can be left empty.

#### Activate SPI master driver

* Modify `config.txt` and remove the comment before `dtparam=spi=on`. [more info](https://www.raspberrypi.org/documentation/hardware/raspberrypi/spi/README.md)
* Add the following to the end of the file:

```ini
dtoverlay=gpio-shutdown
enable_uart=1
```

### Wiring up the RFID module

* The following schema shows the wiring for the RFID-RC522 Mifare reader.
* ATTENTION: Please note that some vendors have a slightly different layout on the module and that the free pin may be at a different position.

![Wiring](src/PiServer/rpi-mfrc522-wiring2.png)

### Connecting to the Raspberry Pi

* Put the SD-Card into the Raspberry Pi
* Connect the Raspberry Pi with the USB-Charger
* Try to find out the [IP of your device](https://www.raspberrypi.org/documentation/remote-access/ip-address.md)
    * If you have multiple Raspberry Pi in the network and want to check if you are SSHing into the right one, you can change the color of the built-in LEDs and see on which Pi it happens: [https://www.jeffgeerling.com/blogs/jeff-geerling/controlling-pwr-act-leds-raspberry-pi](https://www.jeffgeerling.com/blogs/jeff-geerling/controlling-pwr-act-leds-raspberry-pi)
* Connect to the Raspberry Pi via SSH (you can use [putty](https://www.putty.org/) on Windows)
* Username: "pi", Password: "raspberry"
* Change the default password with the command `passwd`
* Change the hostname with `sudo raspi-config`. [more info](https://geek-university.com/raspberry-pi/change-raspberry-pis-hostname/)

### Installing latest firmware

* `sudo apt-get update`
* `sudo apt-get install curl unzip`
* `curl -sSL -o firmware.zip https://audio-hub.azurewebsites.net/api/latestfirmware`
* `unzip firmware.zip -d firmware && rm firmware.zip`
* `chmod u+x firmware/install.sh`
* `sudo firmware/install.sh`

### Checking logs

`tail -f PiServer/publish/logs/clientlog.log`

### Testing self-built Firmware

If you want to run custom firmware on the Raspberry Pi then make changes in the `src/PiServer` project and do the following:

* Make sure the latest version in `RELEASE_NOTES.md` is higher than the latest released version.
* Run `build.cmd CreateFirmware`
* Copy your firmware to the Rasperry Pi - e.g. `pscp -r temp/ pi@192.168.178.28:/home/pi/` (You need to use the IP of your Raspberry Pi)
* SSH into your PI and run `sudo service PiServer restart`. This process will take a while and put your firmware in place.

RELEASE_NOTES.md
# Release Notes

## 1.9.0 - 2019-11.27
* Update dependencies

## 1.8.3 - 2019-10-31
* Use GPIO from .NET Core 3

## 1.7.8 - 2019-10-30
* .NET Core 3

## 1.6.1 - 2019-10-16
* Removed Sonos support
* Key Vault

## 1.5.23 - 2019-08-22
* Sonos support
* Removed Youtube

## 1.4.22 - 2019-08-20
* Update deps
* Use --self-contained for Firmware
* Hook into view

## 1.3.1 - 2019-03-22
* Show light

## 1.2.1 - 2019-03-19
* Update .NET Core to 2.2.3

## 1.1.16 - 2019-03-1
* Upgrade Raspbian

## 1.1.0 - 2019-01-20
* Update from home

## 1.0.15 - 2018-12-28
* Christmas edition

## 0.16.25 - 2018-12-18
* Update to ASP.NET 2.2

## 0.15.11 - 2018-11-05
* Model-Audio-Update

## 0.14.24 - 2018-10-31
* TagHistory websocket

## 0.13.22 - 2018-10-24
* Volume buttons

## 0.12.2 - 2018-10-23
* Volume buttons

## 0.11.21 - 2018-10-23
* Next/Previous buttons

## 0.10.37 - 2018-10-23
* Use better UI

## 0.9.7 - 2018-10-18
* Automatic firmware update

## 0.8.5 - 2018-10-18
* Play youtube playlists

## 0.7.15 - 2018-10-17
* Youtube player

## 0.5.7 - 2018-10-17
* Server logs

## 0.4.4 - 2018-10-16
* logging into logfile

## 0.3.17 - 2018-10-16
* run npm during install

## 0.3.2 - 2018-10-10
* Bundle firmware installer

## 0.2.4 - 2018-10-10
* Stream to Raspberry Pi

## 0.1.12 - 2018-10-08
* Stream from Azure Blob Storage

## 0.0.37 - 2018-10-05
* Initial release


src/PiServer/update.sh
#!/bin/sh

. /lib/lsb/init-functions

SCRIPTROOT=`dirname "$0"`
PROJECTNAME=PiServer
DEFAULTS=/etc/default/$PROJECTNAME
SOURCEPATH=$SCRIPTROOT/bin

if [ ! -e $DEFAULTS -a $SCRIPTROOT/$PROJECTNAME.defaults ]; then
    cp $SCRIPTROOT/$PROJECTNAME.defaults $DEFAULTS
fi

if [ -e $DEFAULTS ]; then
    log_action_begin_msg "Loading $PROJECTNAME settings"
    . /etc/default/$PROJECTNAME
    log_action_end_msg $?

    if [ ! -e $DAEMON ]; then
        log_failure_msg "$PROJECTNAME not installed! Update "
        exit 1
    fi

    if [ -e $SOURCEPATH/$NAME ]; then
        log_action_begin_msg "Stop service"
        service $PROJECTNAME stop
        retval=$?
        log_action_end_msg $retval
        if [ $retval -ne 0 ]; then
            exit 2
        fi

        log_action_begin_msg "Updating packages "
        log_action_cont_msg " apt update "
        apt-get -qq update
        log_action_cont_msg " apt upgrade "
        apt-get -qq upgrade
        log_action_cont_msg " apt install "
        apt-get -qq install  curl libunwind8 gettext apt-transport-https omxplayer wiringpi npm unzip
        log_action_end_msg $?

        log_action_begin_msg "Updating"
        log_action_cont_msg " Init script "
        cp $SCRIPTROOT/$PROJECTNAME /etc/init.d/

        chmod +x /etc/init.d/$PROJECTNAME
        log_action_cont_msg "Reloading system.d"
        systemctl daemon-reload
        log_action_end_msg $?

        log_action_begin_msg "Updating $DAEMONHOME "
        log_action_cont_msg " Backup node_modules "
        cp -a $DAEMONHOME/node_modules $SOURCEPATH/

        log_action_cont_msg " installing new version to $DAEMONHOME "
        rm -R $DAEMONHOME && cp -a $SOURCEPATH/. $DAEMONHOME
        log_action_end_msg $?

        log_action_begin_msg "Updating node_modules"
        cd $DAEMONHOME && npm --silent install > /dev/null 2>&1
        log_action_end_msg $?

        if [ -e $DAEMON ]; then
            chmod +x $DAEMONHOME/$NAME

            log_action_begin_msg "Starting service"
            service $PROJECTNAME start
            log_action_end_msg $?
        fi
    else
        log_failure_msg "Update package corrupt?"
    fi
else
    log_failure_msg "Unable to load defaults"
fi


src/PiServer/install.sh
#!/bin/sh

. /lib/lsb/init-functions

SCRIPTROOT=`dirname "$0"`
SOURCEPATH=$SCRIPTROOT/bin
PROJECTNAME=PiServer
DEFAULTDAEMONHOME=/home/pi/$PROJECTNAME/publish
DEFAULTS=/etc/default/$PROJECTNAME

DAEMONHOME=$1

log_action_begin_msg "Loading $PROJECTNAME settings"
if [ "x$DAEMONHOME" = "x" ]; then
    if [ -e  $SCRIPTROOT/$PROJECTNAME.defaults ]; then
        . $SCRIPTROOT/$PROJECTNAME.defaults
    else
        DAEMONHOME=$DEFAULTDAEMONHOME
    fi
fi
log_action_end_msg $?

log_action_begin_msg "Updating packages "
log_action_cont_msg " apt update "
apt-get -qq update
log_action_cont_msg " apt upgrade "
apt-get -qq upgrade
log_action_cont_msg " apt install "
sudo apt-get install -y curl libunwind8 gettext apt-transport-https omxplayer npm unzip
log_action_end_msg $?

if [ -e  $SCRIPTROOT/$PROJECTNAME ]; then
    if [ ! -e $DEFAULTS -a $SCRIPTROOT/$PROJECTNAME.defaults ]; then
        echo DAEMONHOME=$DAEMONHOME > $DEFAULTS
        cat $SCRIPTROOT/$PROJECTNAME.defaults | grep -v "DAEMONHOME=" >> $DEFAULTS
    fi

    if [ ! -d $DAEMONHOME ]; then
        mkdir -p $DAEMONHOME
    fi

    log_action_begin_msg "Install application into $DAEMONHOME "
    cp -a $SOURCEPATH/. $DAEMONHOME
    cp  $SCRIPTROOT/$PROJECTNAME /etc/init.d/
    log_action_end_msg $?

    log_action_begin_msg "Updating node_modules"
    cd $DAEMONHOME && npm --silent install > /dev/null 2>&1
    log_action_end_msg $?

    log_action_begin_msg "Enabling daemon"
    chmod +x /etc/init.d/$PROJECTNAME
    chmod +x $DAEMONHOME/$NAME
    systemctl daemon-reload
    systemctl enable $PROJECTNAME
    log_action_end_msg $?

    log_action_begin_msg "Starting service"
    service $PROJECTNAME start
    log_action_end_msg $?
fi


src/Client/webpack.config.js
// File adapted from https://github.com/fable-compiler/webpack-config-template
var path = require("path");

function resolve(filePath) {
    return path.join(__dirname, filePath)
}

var CONFIG = {
    fsharpEntry:
        ["whatwg-fetch",
            "@babel/polyfill",
            resolve("./Client.fsproj")
        ],
    outputDir: resolve("./public"),
    devServerProxy: {
        '/api/*': {
            target: 'http://localhost:' + (process.env.SUAVE_FABLE_PORT || "5000"),
            changeOrigin: true
        }
    },
    contentBase: __dirname,
    // Use babel-preset-env to generate JS compatible with most-used browsers.
    // More info at https://github.com/babel/babel/blob/master/packages/babel-preset-env/README.md
    babel: {
        presets: [
            ["@babel/preset-env", {
                "targets": {
                    "browsers": ["last 2 versions"]
                },
                "modules": false,
                "useBuiltIns": "usage",
            }]
        ],
        plugins: ["@babel/plugin-transform-runtime","@babel/plugin-proposal-class-properties"]
    }
}

var isProduction = process.argv.indexOf("-p") >= 0;
console.log("Bundling for " + (isProduction ? "production" : "development") + "...");

var path = require("path");
var webpack = require("webpack");
var MinifyPlugin = require("terser-webpack-plugin");

var commonPlugins = [
];

module.exports = {
    entry: CONFIG.fsharpEntry,
    // NOTE we add a hash to the output file name in production
    // to prevent browser caching if code changes
    output: {
        path: CONFIG.outputDir,
        publicPath: "/public",
        filename: '[name].js'
    },
    resolve: {
        symlinks: false,
    },
    mode: isProduction ? "production" : "development",
    devtool: isProduction ? undefined : "source-map",
    optimization: {
        // Split the code coming from npm packages into a different file.
        // 3rd party dependencies change less often, let the browser cache them.
        splitChunks: {
            cacheGroups: {
                commons: {
                    test: /node_modules/,
                    name: "vendors",
                    chunks: "all"
                }
            }
        },
        minimizer: isProduction ? [new MinifyPlugin()] : []
    },
    // Besides the HtmlPlugin, we use the following plugins:
    // PRODUCTION
    //      - UglifyJSPlugin: Minimize the CSS
    // DEVELOPMENT
    //      - HotModuleReplacementPlugin: Enables hot reloading when code changes without refreshing
    plugins: isProduction ?
        commonPlugins
        : commonPlugins.concat([
            new webpack.HotModuleReplacementPlugin(),
        ]),
    // Configuration for webpack-dev-server
    devServer: {
        proxy: CONFIG.devServerProxy,
        hot: true,
        inline: true,
        historyApiFallback: true,
        contentBase: CONFIG.contentBase
    },
    // - fable-loader: transforms F# into JS
    // - babel-loader: transforms JS to old syntax (compatible with old browsers)
    module: {
        rules: [
            {
                test: /\.fs(x|proj)?$/, 
                use: {
                    loader: "fable-loader",
                    options: {
                        babel: CONFIG.babel,
                        define: isProduction ? [] : ["DEBUG"]
                   }
                },
            },
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: 'babel-loader',
                    options: CONFIG.babel
                },
            },
            {
              test: /\.css$/,
              use: [ 'style-loader', 'css-loader' ]
            },
            {
              test: /\.scss$/,
              use: [ "style-loader", "css-loader", "sass-loader" ]
            }
        ]
    }
};

src/PiServer/read-tag.js
"use strict";
const Mfrc522 = require("mfrc522-rpi");
const SoftSPI = require("rpi-softspi");

const softSPI = new SoftSPI({
    clock: 23, // pin number of SCLK
    mosi: 19, // pin number of MOSI
    miso: 21, // pin number of MISO
    client: 24 // pin number of CS
  });
  
// GPIO 24 can be used for buzzer bin (PIN 18), Reset pin is (PIN 22).
const mfrc522 = new Mfrc522(softSPI).setResetPin(22).setBuzzerPin(18);

function decimalToHex(d) {
    var padding = 2;
    var hex = Number(d).toString(16);
    padding = typeof (padding) === "undefined" || padding === null ? padding = 2 : padding;

    while (hex.length < padding) {
        hex = "0" + hex;
    }

    return hex;
}

exports.read = function (callback, lastTag) {
    mfrc522.reset();

    let response = mfrc522.findCard();
    if (!response.status) {
        callback(null, "");
        return;
    }

    //# Get the UID of the card
    response = mfrc522.getUid();
    if (!response.status) {
        callback(null, "");
        return;
    }

    const uid = response.data;
    var data = uid.map(x => decimalToHex(x)).join('');
    callback(null, data);
}
